server/src/main/java/org/apache/iotdb/db/query/fill/PreviousFill.java

```diff
@@ -59,28 +84,184 @@ public long getBeforeRange() {
+      if (shouldUpdate(lastPointResult.getTimestamp(), lastVersion,
+          lastChunkPoint.getTimestamp(), chunkMetadata.getVersion())) {
+        lastPointResult = lastChunkPoint;
+        lastVersion = chunkMetadata.getVersion();
       }
-      cachedPair = new TimeValuePair(batchData.currentTime(), batchData.currentTsPrimitiveType());
-      batchData.next();
-      if (cachedPair.getTimestamp() <= queryTime) {
-        beforePair = cachedPair;
-      } else {
+    }
+    return lastPointResult;
+  }
+
+  /** Pick up and cache the last sequence TimeseriesMetadata that satisfies timeFilter */
+  private TimeValuePair retrieveValidLastPointFromSeqFiles() throws IOException {
+    List<TsFileResource> seqFileResource = dataSource.getSeqResources();
+    TimeValuePair lastPoint = new TimeValuePair(0, null);
+    for (int index = seqFileResource.size() - 1; index >= 0; index--) {
+      TsFileResource resource = seqFileResource.get(index);
+      TimeseriesMetadata timeseriesMetadata =
+          FileLoaderUtils.loadTimeSeriesMetadata(
+              resource, seriesPath, context, timeFilter, allSensors);
+      if (timeseriesMetadata != null) {
+        if (endtimeContainedByTimeFilter(timeseriesMetadata.getStatistics())) {
+          return constructLastPair(
+              timeseriesMetadata.getStatistics().getEndTime(),
+              timeseriesMetadata.getStatistics().getLastValue(),
+              dataType);
+        } else {
+          List<ChunkMetadata> seqChunkMetadataList =
+              FileLoaderUtils.loadChunkMetadataList(timeseriesMetadata);
+
+          for (int i = seqChunkMetadataList.size() - 1; i >= 0; i--) {
+            lastPoint = getChunkLastPoint(seqChunkMetadataList.get(i));
+            // last point of this sequence chunk is valid, quit the loop
+            if (lastPoint.getValue() != null) {
+              return lastPoint;
+            }
+          }
+        }
+      }
+    }
+
+    return lastPoint;
+  }
+
+  /**
+   * find the last TimeseriesMetadata and unpack all overlapped seq/unseq files
+   */
+  private void UnpackOverlappedUnseqFiles(long lBoundTime) throws IOException {
+    PriorityQueue<TsFileResource> unseqFileResource =
+        sortUnSeqFileResourcesInDecendingOrder(dataSource.getUnseqResources());
+
+    while (!unseqFileResource.isEmpty()) {
+      // The very end time of unseq files is smaller than lBoundTime,
+      // then skip all the rest unseq files
+      if (unseqFileResource.peek().getEndTimeMap().get(seriesPath.getDevice()) < lBoundTime) {
+        return;
+      }
+      TimeseriesMetadata timeseriesMetadata =
+          FileLoaderUtils.loadTimeSeriesMetadata(
+              unseqFileResource.poll(), seriesPath, context, timeFilter, allSensors);
+      if (timeseriesMetadata != null
+          && lBoundTime <= timeseriesMetadata.getStatistics().getEndTime()) {
+        lBoundTime = Math.max(lBoundTime, timeseriesMetadata.getStatistics().getStartTime());
```

Comment:
if queried interval is [0,100]

if seq is [90,100], lBoundTime = 100.

if unseq are [1,10], [1,20], [20,30]. then these unseq files will all be unpacked. This should be avoided